### ISSUE TITLE
Add icon to marker

### ISSUE THREAD

**Issue by francoro** [mapsplugin/cordova-plugin-googlemaps#1614]
In the version 1.4 of the plugin i added the marker like this:

```
let markerOptions = {
      position: latlng,
      disableAutoPan: true,
      icon: {
          url: "./assets/marker-imagen/marker.png",
          size: {
          width: 47,
          height: 60
        }
      }
 };               
const marker: Marker = map.addMarker(markerOptions).then((marker: Marker) => {})
```

Now in the v2 it seems that is not working.
I tried with the setIcon() method this is my code:

```
const marker: Marker = map.addMarker(markerOptions)
                            .then((marker: Marker) => {
                                let iconData: MarkerIcon = "./assets/marker-imagen/marker.png";
                                marker.setIcon(iconData);
})
```
But i have this error in the VS code
Type '"./assets/marker-imagen/marker.png"' has no properties in common with type 'MarkerIcon'.

Do you know how is the correct way to add an icon in the marker?

**Comment by wf9a5m75** [comment]
You can create a marker as the same as v1.

But I haven't checked the ionic wrapper version. There are many mistakes. I have been checking all methods one by one.

**Comment by francoro** [comment]
Okay. I have another issue maybe you know the answer. I enter the view of the map for the first time and shows me gray div but the code with markers is loaded and the html too. but then i go back and reenter again and it works.

**Comment by wf9a5m75** [comment]
Could you take a screen record (video)?

**Comment by francoro** [comment]
Just happen in a old phone now it doesnt happen anymore so no worries

**Comment by wf9a5m75** [comment]
OK, I guess the native API takes longer time than new phone to load the map tiles.

Any other questions?

**Comment by francoro** [comment]
The icon works when i do ionic build android with ionic run android i think it has a different folder for the assets the only problem was that it shows to big and it doenst work the width and heigth:

```
icon: {
    url: "assets/marker-imagen/marker.png",
    size: {
         width: 47,
         height: 60
     }
 }
const marker = map.addMarker(markerOptions)
                            .then((marker) => { })
```
Is there a different way to set the width and height in the v2?

**Comment by wf9a5m75** [comment]
Do you mean the `size` property does not work correctly?

**Comment by francoro** [comment]
Exactly

**Comment by wf9a5m75** [comment]
Okay.
I'm working on another issue right now.
But I will test the size property issue in today. Please wait for a while.
Maybe try tomorrow please (I don't know where you are in the world though).

**Comment by francoro** [comment]
Ok great no hurries, i am from Argentina by the way.

**Comment by wf9a5m75** [comment]
I tested using the JS version with this code, and it worked perfectly.

```js
var map = plugin.google.maps.Map.getMap(div);
map.one(plugin.google.maps.event.MAP_READY, function() {

  map.addMarker({
    'position': {
      lat: 0,
      lng: 0
    },
    icon: {
      url: "../images/number 1-icon.png",
      size: {
        width: 60,
        height: 60
      }
    }
  });
});

```


![screen shot 2017-08-08 at 12 01 16 am](https://user-images.githubusercontent.com/167831/29060026-e815c452-7bcc-11e7-94c2-beb896590d69.png)

**Comment by ryanhalley** [comment]
I am having issues with icon urls this on Android (maybe iOS too), with v2beta 3.
If I create a map with markers, and set the icon url, the map will appear and show the correct icons, but, creating further map instances will then fail. As soon as I comment out the url property, I have no problem creating new map instances; there is a good chance this is related to some general issues with the ionic plugin, as I've raised another ticket to do with creating the third of three map instances not rendering correctly.

**Comment by wf9a5m75** [comment]
@ryanhalley Before posting your question, you should try the latest version.
`cordova-plugin-googlemaps 2.0.0-beta3-20170805-2349` is old.

**Comment by wf9a5m75** [comment]
And you should show me the code. Otherwise, I ignore your comment, because I don't have enough information, so I can not help you.

**Comment by ryanhalley** [comment]
I'll try the latest build, thanks.

                                icon: {
                                    url: 'www/assets/img/map_icon.png'
                                    size: {
                                        width: 10,
                                        height: 12
                                    }
                                },

**Comment by ryanhalley** [comment]
With that enabled (on this old build), I can create the map with the markers, then destroy the page, and if I recreate the page it won't render.

**Comment by ryanhalley** [comment]
Oh, sorry, that is the version I am using.

**Comment by wf9a5m75** [comment]
In v2, you don't need to add `www`

**Comment by wf9a5m75** [comment]
If you want to help me, you need to create an example app. Otherwise, no help.

**Comment by ryanhalley** [comment]
Thanks - I tried the latest build, no change though. I'll work on an example tomorrow, cheers.

**Comment by ryanhalley** [comment]
Hi, I've replicated the issue.

Scenarios:
1. Map with 1 marker not using an icon url: loads fine.
2. Map with 100 marks not using an icon url: loads fine.
3. Map with 1 marker using an icon url; loads fine.
4. Map with 2 or more markers using the same icon url - one of two outcomes occur:
- the map may load correctly and render all markers, but if you leave the page and reload it, the map will not render. All future map loads will not render.
- the may will not load correctly, and you will see only the bottom 1cm of the map, moved to the top of the screen. All future map loads will not render.

import { Component } from '@angular/core';
import {
    GoogleMaps,
    GoogleMap,
    GoogleMapsEvent,
    LatLng,
    CameraPosition,
    MarkerOptions,
    Marker
} from '@ionic-native/google-maps';

@Component({
    templateUrl: 'findMap2.html',
    selector: 'findMap2'
})
export class FindMapPage2 {
    private _map: GoogleMap;
    private _lat: number;
    private _lng: number;

    constructor(
        private googleMaps: GoogleMaps
        // private _navController: NavController,
    ) {

    }

    // Load map only after view is initialized
    ngAfterViewInit() {
        this.loadMap();
    }

    public ionViewWillLeave() {
        this._map.remove();
    }

    loadMap() {
        // create a new map by passing HTMLElement
        let element: HTMLElement = document.getElementById('map');

        this._lat = 50.730502;
        this._lng = -1.984551;

        let cameraPosition: CameraPosition = {
            target: {
                lat: this._lat,
                lng: this._lng
            },
            zoom: 18,
            bearing: 0
        };
        let mapOptions = {
            // This seems to be invalid in v2
            // 'backgroundColor': 'white',
            controls: {
                compass: true,
                myLocationButton: true,
                indoorPicker: true,
                zoom: true
            },
            gestures: {
                scroll: true,
                tilt: true,
                rotate: true,
                zoom: true
            },
            camera: cameraPosition
        };

        this._map = this.googleMaps.create(element, mapOptions);

        // listen to MAP_READY event
        // You must wait for this event to fire before adding something to the map or modifying it in anyway
        this._map.one(GoogleMapsEvent.MAP_READY).then(
            () => {
                console.log('Map is ready!');
                // Now you can add elements to the map like the marker
                this.addMarkers();
            }
        );

    }

    private addMarkers() {
        let lngCurrent = this._lng;
        // Icons only work if there is 1 instance of them
        let numberOfMarkers = 1;

        for (let x = 0; x < numberOfMarkers; x++) {
            // Move each successive marker a little
            lngCurrent += 0.0002;

            let markerOptions: MarkerOptions = {
                position: {
                    lat: this._lat,
                    lng: lngCurrent
                },
                icon: {
                    // Note: If the page does not render correctly, I have to quit: 'ionic cordova run android -cl'
                    // , otherwise, auto-rebuilding on file saves will not render the map correctly in future builds

                    // 1. With no url, the page renders.

                    // 2. Using this url will incorrectly render the map, showing only the bottom 1cm, but moved to the top of the screen
                    url: 'www/assets/img/map_icon_pub.png',

                    // 3. Using this url will render the map once correctly, but if you leave the page and reload it, map is blank.
                    // url: 'www/assets/img/map_icon_club_key_android.png',
                    size: {
                        width: 10,
                        height: 12
                    }
                },
                title: x.toString(),
                snippet: x.toString()
            };

            this._map.addMarker(markerOptions)
                .then(addedMarker => {
                });
        }
    }

}

**Comment by wf9a5m75** [comment]
Please share your code on github

**Comment by wf9a5m75** [comment]
And as I answered before `DO NOT attach www` in the path of icon image file.

**Comment by wf9a5m75** [comment]
And as I describe here, you **should not** use the `ngAfterViewInit`.

```
  constructor(private googleMaps: GoogleMaps, public platform: Platform) {

    // Wait the native plugin is ready.
    platform.ready().then(() => {
      this.loadMap();
    });
  }

  // Don't use the ngAfterViewInit(). The native plugin is not ready.
  //ngAfterViewInit() {
  // this.loadMap();
  //}
```
https://github.com/mapsplugin/cordova-plugin-googlemaps-doc/tree/master/v2.0.0/ionic-native#usage

Or just simply wrap with setTimeout()

```
let self = this;
setTimeout(function() {
  self.loadMap();
}, 500);
```

**Comment by ryanhalley** [comment]
Here are my plugins again, in case they have changed:
com.telerik.plugins.nativepagetransitions 0.6.5 "Native Page Transitions"
com.xmartlabs.cordova.market 1.1 "Market"
cordova-custom-config 3.2.0 "cordova-custom-config"
cordova-plugin-actionsheet 2.3.3 "ActionSheet"
cordova-plugin-app-name 1.0.0 "App Namer"
cordova-plugin-app-version 0.1.9 "AppVersion"
cordova-plugin-camera 2.4.1 "Camera"
cordova-plugin-compat 1.1.0 "Compat"
cordova-plugin-console 1.0.6 "Console"
cordova-plugin-contacts 2.3.0 "Contacts"
cordova-plugin-device 1.1.5 "Device"
cordova-plugin-email 1.2.6 "EmailComposer"
cordova-plugin-geolocation 2.4.2 "Geolocation"
cordova-plugin-google-analytics 1.6.0 "Google Universal Analytics Plugin"
cordova-plugin-googlemaps 2.0.0-beta3-20170808-2240 "cordova-plugin-googlemaps"
cordova-plugin-inapppurchase 1.1.0 "In App Purchase"
cordova-plugin-network-information 1.3.2 "Network Information"
cordova-plugin-splashscreen 4.0.2 "Splashscreen"
cordova-plugin-statusbar 2.2.2 "StatusBar"
cordova-plugin-vibration 2.1.5 "Vibration"
cordova-plugin-whitelist 1.3.2 "Whitelist"
cordova-plugin-x-socialsharing 5.1.6 "SocialSharing"
cordova-sqlite-storage 1.5.3 "Cordova sqlite storage plugin"
ionic-plugin-keyboard 2.2.1 "Keyboard"
phonegap-plugin-push 1.10.5 "PushPlugin"
uk.co.workingedge.phonegap.plugin.launchnavigator 3.2.2 "Launch Navigator"

**Comment by wf9a5m75** [comment]
Okay, please share your **project files (not HTML and image files)** on **github**.
Please(x3) do not send as zip file.

**Comment by ryanhalley** [comment]
OK, thanks, will do.

**Comment by wf9a5m75** [comment]
Thank you.

**Comment by ryanhalley** [comment]
Hi, project added here: https://github.com/ryanhalley/googlemapsiconurl.git

**Comment by wf9a5m75** [comment]
@ryanhalley Let me confirm one thing, do you run your app with `$> ionic cordova run android(or ios) -l`, don't you?

**Comment by ryanhalley** [comment]
ionic cordova run android -cl

**Comment by wf9a5m75** [comment]
@ryanhalley I fixed both bugs,
- map is not displayed after coming back from another page,
- and icon is not displayed especially on live loading.

Please reinstall the latest version.

**Comment by wf9a5m75** [comment]
By the way, changing the background color is 
https://github.com/mapsplugin/cordova-plugin-googlemaps-doc/blob/master/v2.0.0/class/Environment/setBackgroundColor/README.md

**Comment by ryanhalley** [comment]
Thank you - both bugs confirmed as fixed with livereload, and in release builds!
Also, thanks for the setBackgroundColor tip!

There is one last niggling glitch relating to livereload only which I'll mention, since you've already touched on it. If you are using livereload and you have a map open, and then save a change to the code, when the app reloads it fails to render the first created map about 50% of the time, until you leave the page and reopen it, (until your latest fix, that previously would not fix it, but now it does). It's as if there is some cleanup which needs to be done during the reload, similar to the fix you added. If are you able to make it always work immediately after a livereload, then the development process will be 100% painless.

Finally, you have been very helpful, thank you very much for these fixes!

**Comment by francoro** [comment]
You wf9a5m75 put mi this above. i dont undestand. You can mix the JS version with the native?
The JS version is when start with plugin.google.maps right? I can use that in angular 4 and VS code? because the type checker of typescript didnt let you put plugin.google.maps without import something.


I tested using the JS version with this code, and it worked perfectly.

var map = plugin.google.maps.Map.getMap(div);
map.one(plugin.google.maps.event.MAP_READY, function() {

  ```
map.addMarker({
    'position': {
      lat: 0,
      lng: 0
    },
    icon: {
      url: "../images/number 1-icon.png",
      size: {
        width: 60,
        height: 60
      }
    }
  });
});
```

**Comment by wf9a5m75** [comment]
https://github.com/mapsplugin/cordova-plugin-googlemaps-doc/blob/master/v2.0.0/ionic-native/README.md

**Comment by francoro** [comment]
The example in that link doesn't use icon: { size: {width;, height:}}. i am triyng without the ionic wrapper and still shows the icons to Big giants would say. This is my code:

let markerOptions = {
    position: latlng,
    data: locations[i].id,
    disableAutoPan: true,
    icon: {
          url: "assets/marker-imagen/marker.png",
          size: {
             width: 47,
             height: 60
          }
      }                 
};
const marker = map.addMarker(markerOptions)
                            .then((marker) => {
                            })

**Comment by wf9a5m75** [comment]
![1614](https://user-images.githubusercontent.com/167831/29222449-5f4d9b2e-7e77-11e7-8173-a1fa48d572e2.gif)

**Comment by wf9a5m75** [comment]
Here is iOS demo.
![1614](https://user-images.githubusercontent.com/167831/29223120-3f949ff0-7e7a-11e7-9c57-c780a7c107da.gif)

**Comment by francoro** [comment]
Its working now. Thanks

**Comment by chrillewoodz** [comment]
I too must add `www` at the beginning or it simply doesn't work.

**Comment by JayanthkumarJK** [comment]
this code is working perfectly for me
```let targetLocation = new LatLng("your latitute", "your longitude");

    //code for marker and icon 

    let markerOption: MarkerOptions = {
      position: targetLocation,
      title: "if u want any title to display",  
      icon: {
        url: "ex: assests/img/youricon",
        size: {
            width: 10"based on your size",
            height: 12"based on your requirement"
        }
        }
    };```

dont forget to add ionic native plugin
import {GoogleMaps,GoogleMap,GoogleMapsEvent,LatLng,CameraPosition,MarkerOptions,Marker, MarkerIcon} from '@ionic-native/google-maps'; 

Thankyou
